src/server/status_server.rs

```diff
@@ -41,26 +87,104 @@ impl StatusServer {
         }
     }
 
+    pub fn dump_prof(seconds: u64) -> Box<dyn Future<Item = Vec<u8>, Error = ProfError> + Send> {
+        let lock = match profiler_guard::ProfLock::new() {
+            Err(e) => return Box::new(err(e)),
+            Ok(lock) => lock,
+        };
+        info!("Start profiling {} seconds", seconds);
+
+        let timer = GLOBAL_TIMER_HANDLE.clone();
+        Box::new(lock.then(move |guard| {
+            timer
+                .delay(std::time::Instant::now() + std::time::Duration::from_secs(seconds))
+                .then(move |_| {
+                    let tmp_dir = TempDir::new("").unwrap();
+                    let os_path = tmp_dir.path().join("tikv_dump_profile").into_os_string();
+                    let path = os_path.into_string().unwrap();
+
+                    tikv_alloc::dump_prof(Some(&path));
+                    drop(guard);
+                    tokio_fs::file::File::open(path)
+                        .and_then(|file| {
+                            let buf: Vec<u8> = Vec::new();
+                            tokio_io::io::read_to_end(file, buf)
+                        })
+                        .and_then(move |(_, buf)| {
+                            drop(tmp_dir);
+                            ok(buf)
+                        })
+                        .map_err(|e| -> ProfError { e.into() })
+                })
+        }))
+    }
+
+    pub fn dump_prof_to_resp(
+        req: Request<Body>,
+    ) -> Box<dyn Future<Item = Response<Body>, Error = hyper::Error> + Send> {
+        let url = url::Url::parse(&format!("http://host{}", req.uri().to_string())).unwrap(); // Add scheme and host to parse query
+        let query_pairs: HashMap<_, _> = url.query_pairs().collect();
+        let seconds: u64 = match query_pairs.get("seconds") {
+            Some(val) => match val.parse() {
+                Ok(val) => val,
+                Err(_) => {
+                    let response = Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::empty())
+                        .unwrap();
+                    return Box::new(ok(response));
+                }
+            },
+            None => {
+                let response = Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .body(Body::empty())
+                    .unwrap();
+                return Box::new(ok(response));
+            }
+        };
+
+        Box::new(
+            Self::dump_prof(seconds)
+                .and_then(|buf| {
+                    let response = Response::builder().body(buf.into()).unwrap();
```

Comment:
Please set the "content-length" and "content-type" headers. The header constants are in `http::header` and the mime type is in the `mime` crate. I don't _think_ the response builder will automatically set the "content-length" from the body.